paddle/fluid/imperative/jit/program_desc_tracer.cc

```diff
@@ -221,40 +220,47 @@ void ProgramDescTracer::InsertVarIfNotExist(
     const std::shared_ptr<VarBase> &new_var, bool is_input) {
   PADDLE_ENFORCE_NOT_NULL(new_var, platform::errors::InvalidArgument(
                                        "The variable to insert is NULL."));
-  if (vars_.count(new_var) != 0) return;
 
-  auto new_var_desc = new framework::VarDesc("");
-  vars_[new_var].reset(new_var_desc);
+  for (; vars_.count(new_var) == 0;) {
+    if (!vars_mutex_.try_lock()) continue;
```

Comment:
Some questions:

- When would the `ProgramDescTracer` run in multiple threads? Users usually use imperative programming in a single thread.
- It is still not thread safe after this PR. There may be 2 threads running in the same time: thread 1 runs `vars_.count(new_var)` and thread 2 runs `vars_[new_var]`. There is still data race.

Reply:
Thanks for your comments ;) 
On the second point first. The C++ standard stipulates that a const-member function of a container class should be thread-safe by default. I suppose std:: implementation follows the standard.  
Regarding the first question, there seem to be quite a few places in `imperative.cc` where the python GIL is released right before calling `tracer->traceop`. I'm not sure why multithreading happens in the imperative mode in the first place, it seems so...

Reply:
Some background information. The issue was raised during an 8-GPU training and the triggering point is in ProgramDescTracer::InsertVarIfNotExist..

Reply:
Agree. Both `vars_.count()` and `vars_[]` should be protected..